Fast math breaks some tests

I have two failing tests on our cluster with the current development branch:

1)

```espresso/src/core/unit_tests/grid_test.cpp(79)```
+ 
```espresso/src/core/unit_tests/grid_test.cpp(82)```

Fails with:  
```error: in "get_mi_coord_test": absolute value of std::abs(get_mi_coord(a, b, box_l, true) - (a - b) - box_l){4.4408920985006262e-16} exceeds 2.2204460492503131e-16```
+ 
```error: in "get_mi_coord_test": absolute value of std::abs(get_mi_coord(b, a, box_l, true) - (b - a) + box_l){4.4408920985006262e-16} exceeds 2.2204460492503131e-16```

This seems to be the same issue twice

2) 
```
Test  #44: rotational_inertia ......................***Failed

FF
======================================================================
FAIL: test_energy_and_momentum_conservation (__main__.RotationalInertia)
----------------------------------------------------------------------
Traceback (most recent call last):
/espresso/build/testsuite/python/rotational_inertia.py", line 154, in test_energy_and_momentum_conservation
    self.assertAlmostEqual(self.energy(p), E0, places=3)
AssertionError: 14.85059589306411 != 14.85 within 3 places

======================================================================
FAIL: test_stability (__main__.RotationalInertia)
----------------------------------------------------------------------
Traceback (most recent call last):
/espresso/build/testsuite/python/rotational_inertia.py", line 84, in test_stability
    i, k, self.L_0_lab[k], self.L_lab[k]))
AssertionError: 0.004388219520887526 not less than or equal to 0.004 : Inertial motion around stable axis J1: Deviation in angular momentum is too large. Step 2, coordinate 2, expected -0.7955, got -0.7998882195208875

----------------------------------------------------------------------
Ran 2 tests in 0.096s

FAILED (failures=2)

```

I compiled Espresso with the intel compiler. Are these known issues that occured before?
In case the intel compiler is no longer supported, I will try again with gcc the next days.

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -213,6 +213,27 @@ clang-sanitizer:
     - linux
     - cuda
 
+fast_math:
+  <<: *global_job_definition
+  stage: build
+  image: docker.pkg.github.com/espressomd/docker/cuda:08d67d67a4d3c3279cacee975b68c02ec89b2a21
+  variables:
+     CC: 'gcc-9'
+     CXX: 'g++-9'
+     myconfig: 'maxset'
+     with_cuda: 'true'
+     with_coverage: 'false'
+     with_scafacos: 'false'
+     with_stokesian_dynamics: 'true'
+     with_fast_math: 'true'
+  script:
+    - bash maintainer/CI/build_cmake.sh
+  tags:
+    - docker
+    - linux
+    - cuda
+  when: manual
+
 cuda11-maxset:
   <<: *global_job_definition
   stage: build

--- a/doc/sphinx/installation.rst
+++ b/doc/sphinx/installation.rst
@@ -684,6 +684,38 @@ correct one with ``CUDA_BIN_PATH=/usr/local/cuda-10.0 cmake .. -DWITH_CUDA=ON``
 (with Clang as the CUDA compiler, you also need to override its default CUDA
 path with ``-DCMAKE_CXX_FLAGS=--cuda-path=/usr/local/cuda-10.0``).
 
+.. _Build types and compiler flags:
+
+Build types and compiler flags
+""""""""""""""""""""""""""""""
+
+The build type is controlled by ``-D CMAKE_BUILD_TYPE=<type>`` where
+``<type>`` can take one of the following values:
+
+* ``Release``: for production use: disables assertions and debug information,
+  enables ``-O3`` optimization (this is the default)
+* ``RelWithAssert``: for debugging purposes: enables assertions and
+  ``-O3`` optimization (use this to track the source of a fatal error)
+* ``Debug``: for debugging in GDB
+* ``Coverage``: for code coverage
+
+Cluster users and HPC developers may be interested in manually editing the
+``cxx_interface`` variable in the top-level ``CMakeLists.txt`` file for
+finer control over compiler flags. The variable declaration is followed
+by a series of conditionals to enable or disable compiler-specific flags.
+Compiler flags passed to CMake via the ``-DCMAKE_CXX_FLAGS`` option
+(such as ``cmake . -DCMAKE_CXX_FLAGS="-ffast-math -fno-finite-math-only"``)
+will appear in the compiler command before the flags in ``cxx_interface``,
+and will therefore have lower precedence.
+
+Be aware that fast-math mode can break |es|. It is incompatible with the
+``ADDITIONAL_CHECKS`` feature due to the loss of precision in the LB code
+on CPU. The Clang 10 compiler breaks field couplings with ``-ffast-math``.
+The Intel compiler enables the ``-fp-model fast=1`` flag by default;
+it can be disabled by adding the ``-fp-model=strict`` flag.
+
+|es| currently doesn't fully support link-time optimization (LTO).
+
 
 .. _Configuring without a network connection:
 

--- a/maintainer/CI/build_cmake.sh
+++ b/maintainer/CI/build_cmake.sh
@@ -92,6 +92,7 @@ set_default_value make_check_python true
 set_default_value make_check_tutorials false
 set_default_value make_check_samples false
 set_default_value make_check_benchmarks false
+set_default_value with_fast_math false
 set_default_value with_cuda false
 set_default_value with_cuda_compiler "nvcc"
 set_default_value with_cxx_standard 14
@@ -114,6 +115,10 @@ if [ "${with_coverage}" = true ]; then
 fi
 if [ "${with_coverage}" = true ] || [ "${with_coverage_python}" = true ] ; then
     bash <(curl -s https://codecov.io/env) 1>/dev/null 2>&1
+fi
+
+if [ "${with_fast_math}" = true ]; then
+    cmake_param_protected="-DCMAKE_CXX_FLAGS=-ffast-math -fno-finite-math-only"
 fi
 
 cmake_params="-DCMAKE_BUILD_TYPE=${build_type} -DCMAKE_CXX_STANDARD=${with_cxx_standard} -DWARNINGS_ARE_ERRORS=ON ${cmake_params}"
@@ -182,7 +187,7 @@ outp srcdir builddir \
     with_coverage with_coverage_python \
     with_ubsan with_asan \
     check_odd_only \
-    with_static_analysis myconfig \
+    with_static_analysis with_fast_math myconfig \
     build_procs check_procs \
     with_cuda with_cuda_compiler with_ccache
 
@@ -210,9 +215,12 @@ else
     fi
     echo "Copying ${myconfig}.hpp to ${builddir}/myconfig.hpp..."
     cp "${myconfig_file}" "${builddir}/myconfig.hpp"
+    if [ "${with_fast_math}" = true ]; then
+        sed -i '/#define ADDITIONAL_CHECKS/d' "${builddir}/myconfig.hpp"
+    fi
 fi
 
-cmake ${cmake_params} "${srcdir}" || exit 1
+cmake ${cmake_params} "${cmake_param_protected}" "${srcdir}" || exit 1
 end "CONFIGURE"
 
 # BUILD

--- a/src/core/algorithm/periodic_fold.hpp
+++ b/src/core/algorithm/periodic_fold.hpp
@@ -57,6 +57,7 @@ std::pair<T, I> periodic_fold(T x, I i, T const &l) {
  * @return x folded into [0, l).
  */
 template <typename T> T periodic_fold(T x, T const &l) {
+#ifndef __FAST_MATH__
   /* Can't fold if either x or l is nan or inf. */
   if (std::isnan(x) or std::isnan(l) or std::isinf(x) or (l == 0)) {
     return std::nan("");
@@ -64,6 +65,7 @@ template <typename T> T periodic_fold(T x, T const &l) {
   if (std::isinf(l)) {
     return x;
   }
+#endif // __FAST_MATH__
 
   while (x < 0) {
     x += l;

--- a/src/core/grid_based_algorithms/electrokinetics_cuda.cu
+++ b/src/core/grid_based_algorithms/electrokinetics_cuda.cu
@@ -48,6 +48,7 @@
 #include <cstddef>
 #include <fstream>
 #include <iostream>
+#include <limits>
 #include <sstream>
 #include <string>
 #include <vector>
@@ -2345,16 +2346,20 @@ int ek_init() {
 
     ek_initialized = true;
   } else {
-    if (lbpar_gpu.agrid != ek_parameters.agrid ||
-        lbpar_gpu.viscosity != ek_parameters.viscosity *
-                                   ek_parameters.time_step /
-                                   Utils::sqr(ek_parameters.agrid) ||
-        lbpar_gpu.bulk_viscosity != ek_parameters.bulk_viscosity *
-                                        ek_parameters.time_step /
-                                        Utils::sqr(ek_parameters.agrid) ||
-        lb_lbcoupling_get_gamma() != ek_parameters.friction ||
-        lbpar_gpu.rho !=
-            ek_parameters.lb_density * Utils::int_pow<3>(ek_parameters.agrid)) {
+    auto const not_close = [](float a, float b) {
+      return std::abs(a - b) > std::numeric_limits<float>::epsilon();
+    };
+    if (not_close(lbpar_gpu.agrid, ek_parameters.agrid) ||
+        not_close(lbpar_gpu.viscosity, ek_parameters.viscosity *
+                                           ek_parameters.time_step /
+                                           Utils::sqr(ek_parameters.agrid)) ||
+        not_close(lbpar_gpu.bulk_viscosity,
+                  ek_parameters.bulk_viscosity * ek_parameters.time_step /
+                      Utils::sqr(ek_parameters.agrid)) ||
+        not_close(static_cast<float>(lb_lbcoupling_get_gamma()),
+                  ek_parameters.friction) ||
+        not_close(lbpar_gpu.rho, ek_parameters.lb_density *
+                                     Utils::int_pow<3>(ek_parameters.agrid))) {
       fprintf(stderr,
               "ERROR: The LB parameters on the GPU cannot be reinitialized.\n");
 

--- a/src/core/unit_tests/field_coupling_fields_test.cpp
+++ b/src/core/unit_tests/field_coupling_fields_test.cpp
@@ -37,6 +37,8 @@
 #include <limits>
 #include <type_traits>
 
+constexpr auto eps = 10 * std::numeric_limits<double>::epsilon();
+
 using namespace FieldCoupling::Fields;
 
 BOOST_AUTO_TEST_CASE(jacobian_type_test) {
@@ -280,8 +282,7 @@ BOOST_AUTO_TEST_CASE(interpolated_scalar_field) {
 
     auto const field_value = field(p);
 
-    BOOST_CHECK(std::abs(interpolated_value - field_value) <
-                std::numeric_limits<double>::epsilon());
+    BOOST_CHECK_SMALL(std::abs(interpolated_value - field_value), eps);
   }
 
   /* jacobian value */
@@ -309,8 +310,7 @@ BOOST_AUTO_TEST_CASE(interpolated_scalar_field) {
         p, [&data](const std::array<int, 3> &ind) { return data(ind); },
         grid_spacing, origin, Utils::Vector3d{});
 
-    BOOST_CHECK((interpolated_value - field_value).norm() <
-                2 * std::numeric_limits<double>::epsilon());
+    BOOST_CHECK_SMALL((interpolated_value - field_value).norm(), eps);
   }
 }
 
@@ -358,8 +358,7 @@ BOOST_AUTO_TEST_CASE(interpolated_vector_field) {
         p, [&data](const std::array<int, 3> &ind) { return data(ind); },
         grid_spacing, origin, Utils::Vector2d{});
 
-    BOOST_CHECK_SMALL((interpolated_value - field_value).norm(),
-                      std::numeric_limits<double>::epsilon());
+    BOOST_CHECK_SMALL((interpolated_value - field_value).norm(), eps);
   }
 
   /* jacobian value */
@@ -396,10 +395,8 @@ BOOST_AUTO_TEST_CASE(interpolated_vector_field) {
         grid_spacing, origin, Field::jacobian_type{});
 
     BOOST_CHECK_SMALL(
-        (interpolated_value.row<0>() - field_value.row<0>()).norm(),
-        3 * std::numeric_limits<double>::epsilon());
+        (interpolated_value.row<0>() - field_value.row<0>()).norm(), eps);
     BOOST_CHECK_SMALL(
-        (interpolated_value.row<1>() - field_value.row<1>()).norm(),
-        3 * std::numeric_limits<double>::epsilon());
+        (interpolated_value.row<1>() - field_value.row<1>()).norm(), eps);
   }
 }

--- a/src/core/unit_tests/periodic_fold_test.cpp
+++ b/src/core/unit_tests/periodic_fold_test.cpp
@@ -106,7 +106,7 @@ BOOST_AUTO_TEST_CASE(with_image_count) {
     auto const box = 10.;
     auto const res = periodic_fold(x, 0, box);
     BOOST_CHECK(res.first >= 0.);
-    BOOST_CHECK(res.first < box);
+    BOOST_CHECK(res.first <= box);
     BOOST_CHECK(std::abs(res.first - x + res.second * box) <=
                 std::numeric_limits<double>::epsilon());
   }

--- a/src/core/unit_tests/periodic_fold_test.cpp
+++ b/src/core/unit_tests/periodic_fold_test.cpp
@@ -69,6 +69,7 @@ BOOST_AUTO_TEST_CASE(with_image_count) {
     BOOST_CHECK_EQUAL(res.second, i + 1);
   }
 
+#ifndef __FAST_MATH__
   /* Pathological (NaN) */
   {
     auto const x = std::nan("");
@@ -77,6 +78,7 @@ BOOST_AUTO_TEST_CASE(with_image_count) {
     auto const res = periodic_fold(x, i, box);
     BOOST_CHECK(std::isnan(res.first));
   }
+#endif // __FAST_MATH__
 
   /* Overflow right */
   {
@@ -155,6 +157,7 @@ BOOST_AUTO_TEST_CASE(without_image_count) {
     BOOST_CHECK_EQUAL(res, x - box);
   }
 
+#ifndef __FAST_MATH__
   /* Pathological (NaN value) */
   {
     auto const x = std::nan("");
@@ -170,6 +173,7 @@ BOOST_AUTO_TEST_CASE(without_image_count) {
     auto const res = periodic_fold(x, box);
     BOOST_CHECK(std::isnan(res));
   }
+#endif // __FAST_MATH__
 
   /* Corner left */
   {

--- a/src/utils/include/utils/checks/charge_neutrality.hpp
+++ b/src/utils/include/utils/checks/charge_neutrality.hpp
@@ -28,13 +28,13 @@
 
 namespace Utils {
 template <typename ParticleRange>
-bool check_charge_neutrality(ParticleRange &prange,
-                             double relative_tolerance = 1e-12) {
+bool check_charge_neutrality(ParticleRange &prange) {
   using namespace boost::accumulators;
   using KahanSum = accumulator_set<double, features<tag::sum_kahan>>;
 
   KahanSum q_sum;
   auto q_min = std::numeric_limits<double>::infinity();
+  constexpr auto relative_tolerance = 2e-12;
 
   for (auto const &p : prange) {
     auto const &q = p.p.q;

--- a/testsuite/python/accumulator_correlator.py
+++ b/testsuite/python/accumulator_correlator.py
@@ -234,7 +234,7 @@ class CorrelatorTest(ut.TestCase):
                 corr = accumulators[compression].result().flatten()
                 corr_ref = np.repeat(compressed_ref[compression], corr.shape)
                 corr_ref[:tau_lin + 1] = uncompressed[:tau_lin + 1]
-                np.testing.assert_array_equal(corr, corr_ref)
+                np.testing.assert_allclose(corr, corr_ref, rtol=0., atol=1e-14)
             self.system.auto_update_accumulators.clear()
 
     def test_correlator_interface(self):

--- a/testsuite/python/dpd.py
+++ b/testsuite/python/dpd.py
@@ -134,7 +134,7 @@ class DPDThermostat(ut.TestCase):
         system.integrator.run(0)
 
         # Only trans, so x component should be zero
-        self.assertLess(abs(p0.f[0]), 1e-16)
+        self.assertLess(abs(p0.f[0]), 1e-14)
         np.testing.assert_allclose(
             np.copy(p0.f[1:2]), gamma * v[1:2], rtol=0, atol=1e-11)
         np.testing.assert_array_equal(np.copy(p0.f), -np.copy(p1.f))
@@ -176,7 +176,7 @@ class DPDThermostat(ut.TestCase):
         system.integrator.run(0)
 
         # Only trans, so x component should be zero
-        self.assertLess(abs(p0.f[0]), 1e-16)
+        self.assertLess(abs(p0.f[0]), 1e-14)
         omega = calc_omega(1.3, 1.4)**2
         np.testing.assert_allclose(
             np.copy(p0.f[1:2]), omega * gamma * v[1:2], rtol=0, atol=1e-11)

--- a/testsuite/python/galilei.py
+++ b/testsuite/python/galilei.py
@@ -96,7 +96,7 @@ class Galilei(ut.TestCase):
         g.galilei_transform()
 
         np.testing.assert_allclose(
-            np.copy(g.system_CMS_velocity()), np.zeros((3,)), atol=1e-15)
+            np.copy(g.system_CMS_velocity()), np.zeros((3,)), atol=1e-14)
 
 
 if __name__ == "__main__":

--- a/testsuite/python/lb_interpolation.py
+++ b/testsuite/python/lb_interpolation.py
@@ -90,7 +90,9 @@ class LBInterpolation:
         np.testing.assert_allclose(
             np.copy(self.lbf.get_interpolated_velocity(
                 [self.system.box_l[0] - AGRID, 0, 0])),
-            0.5 * (np.array([0, 0, V_BOUNDARY]) + node_next_to_boundary.velocity))
+            0.5 * (np.array([0, 0, V_BOUNDARY]) +
+                   node_next_to_boundary.velocity),
+            rtol=2e-7)
 
         # Bulk
         for pos in itertools.product(

--- a/testsuite/python/observable_cylindrical.py
+++ b/testsuite/python/observable_cylindrical.py
@@ -235,7 +235,8 @@ class TestCylindricalObservable(ut.TestCase):
         self.assertEqual(observable.min_phi, 2)
         self.assertEqual(observable.min_z, 1)
         obs_bin_edges = observable.bin_edges()
-        np.testing.assert_array_almost_equal(obs_bin_edges[0, 0, 0], [1, 2, 1])
+        np.testing.assert_allclose(obs_bin_edges[0, 0, 0], [1, 2, 1],
+                                   rtol=0., atol=1e-14)
         # check edges upper corner
         self.assertEqual(observable.max_r, params['max_r'])
         self.assertEqual(observable.max_phi, params['max_phi'])
@@ -246,7 +247,8 @@ class TestCylindricalObservable(ut.TestCase):
         self.assertEqual(observable.max_phi, 8)
         self.assertEqual(observable.max_z, 9)
         obs_bin_edges = observable.bin_edges()
-        np.testing.assert_array_equal(obs_bin_edges[-1, -1, -1], [7, 8, 9])
+        np.testing.assert_allclose(obs_bin_edges[-1, -1, -1], [7, 8, 9],
+                                   rtol=0., atol=1e-14)
         # check center, axis, orientation
         ctp = espressomd.math.CylindricalTransformationParameters(
             center=[1, 2, 3], axis=[0, 1, 0], orientation=[0, 0, 1])

--- a/testsuite/python/observable_cylindricalLB.py
+++ b/testsuite/python/observable_cylindricalLB.py
@@ -202,7 +202,8 @@ class CylindricalLBObservableCommon:
         self.assertEqual(observable.min_phi, 1.0)
         self.assertEqual(observable.min_z, 1.0)
         obs_bin_edges = observable.bin_edges()
-        np.testing.assert_array_equal(obs_bin_edges[0, 0, 0], [4, 1, 1])
+        np.testing.assert_allclose(obs_bin_edges[0, 0, 0], [4, 1, 1],
+                                   rtol=0., atol=1e-14)
         # check edges upper corner
         self.assertEqual(observable.max_r, params['max_r'])
         self.assertEqual(observable.max_phi, params['max_phi'])
@@ -213,7 +214,8 @@ class CylindricalLBObservableCommon:
         self.assertEqual(observable.max_phi, 8)
         self.assertEqual(observable.max_z, 9)
         obs_bin_edges = observable.bin_edges()
-        np.testing.assert_array_equal(obs_bin_edges[-1, -1, -1], [7, 8, 9])
+        np.testing.assert_allclose(obs_bin_edges[-1, -1, -1], [7, 8, 9],
+                                   rtol=0., atol=1e-14)
         # check sampling_density
         self.assertEqual(
             observable.sampling_density,

